OpenShiftClient cannot replace existing resource with API version =! v1

**openshift version:** 3.11
**appeared since:** 4.11.0
**affected objects:** any object that has an API version different than v1, imagestream (image.openshift.io/v1), route (route.openshift.io/v1), ...

It looks like the replace mechanism has changed and you now try to POST an object to catch the error 410 to detect if it already exists. I think the POST somehow sends the wrong API version, compared to what OCP is expecting.

**code sample:**
```
ImageStream imageStream = Serialization.unmarshal(objectAsJsonString, ImageStream.class);
client.imageStreams().inNamespace(namespace).createOrReplace(imageStream);
```

**stacktrace:**
```
2020-08-31 16:42:22,507 ERROR com.mycompany.ocpdeploy.Launcher deploy job failed
java.lang.RuntimeException: java.util.concurrent.ExecutionException: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://ocpdev.dev.biz.lodh.com:8443/apis/image.openshift.io/v1/namespaces/ocptest-kblg-kuk/imagestreams. Message: the API version in the data (image.openshift.io/v1) does not match the expected API version (v1). Received status: Status(apiVersion=v1, code=400, details=null, kind=Status, message=the API version in the data (image.openshift.io/v1) does not match the expected API version (v1), metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=BadRequest, status=Failure, additionalProperties={}).
	at com.mycompany.ocpdeploy.deploy.OpenShiftManager.ocApply(OpenShiftManager.java:127)
	at com.mycompany.ocpdeploy.deploy.K8sStep.lambda$deployApplicationContext$0(K8sStep.java:147)
	at com.mycompany.ocpdeploy.deploy.helper.TaskHelper.withRetries(TaskHelper.java:38)
	at com.mycompany.ocpdeploy.deploy.K8sStep.deployApplicationContext(K8sStep.java:143)
	at com.mycompany.ocpdeploy.deploy.K8sStep.createAppResources(K8sStep.java:91)
	at com.mycompany.ocpdeploy.deploy.K8sStep.execute(K8sStep.java:77)
	at com.mycompany.ocpdeploy.deploy.DeployJob.deployBase(DeployJob.java:246)
	at com.mycompany.ocpdeploy.deploy.SmbManager.deploy(SmbManager.java:56)
	at com.mycompany.ocpdeploy.deploy.DeployJob.deploy(DeployJob.java:220)
	at com.mycompany.ocpdeploy.deploy.DeployJob.deploy(DeployJob.java:125)
	at com.mycompany.ocpdeploy.deploy.DeployJob.execute(DeployJob.java:86)
	at com.mycompany.ocpdeploy.Launcher.main(Launcher.java:51)
Caused by: java.util.concurrent.ExecutionException: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://ocpdev.dev.biz.lodh.com:8443/apis/image.openshift.io/v1/namespaces/ocptest-kblg-kuk/imagestreams. Message: the API version in the data (image.openshift.io/v1) does not match the expected API version (v1). Received status: Status(apiVersion=v1, code=400, details=null, kind=Status, message=the API version in the data (image.openshift.io/v1) does not match the expected API version (v1), metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=BadRequest, status=Failure, additionalProperties={}).
	at java.base/java.util.concurrent.FutureTask.report(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.get(Unknown Source)
	at com.mycompany.ocpdeploy.deploy.OpenShiftManager.ocApply(OpenShiftManager.java:121)
	... 11 common frames omitted
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://ocpdev.dev.biz.lodh.com:8443/apis/image.openshift.io/v1/namespaces/ocptest-kblg-kuk/imagestreams. Message: the API version in the data (image.openshift.io/v1) does not match the expected API version (v1). Received status: Status(apiVersion=v1, code=400, details=null, kind=Status, message=the API version in the data (image.openshift.io/v1) does not match the expected API version (v1), metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=BadRequest, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:589)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:528)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:492)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:451)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleCreate(OperationSupport.java:252)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleCreate(BaseOperation.java:867)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:356)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.createOrReplace(BaseOperation.java:421)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.createOrReplace(BaseOperation.java:83)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.createOrReplace(BaseOperation.java:415)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.createOrReplace(BaseOperation.java:83)
	at com.mycompany.ocpdeploy.deploy.OpenShiftManager.createOrReplace(OpenShiftManager.java:271)
	at com.mycompany.ocpdeploy.deploy.OpenShiftManager.lambda$ocApply$1(OpenShiftManager.java:114)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,7 @@
 
 #### Bugs
 * Fix: #2442: Wrong resource kind in `ProjectRequestHandler` causes ClassCastException when handling Project resources.
+* Fix #2467: OpenShiftClient cannot replace existing resource with API version =! v1
 
 #### Improvements
 * Fix #2473: Removed unused ValidationMessages.properties

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/BackwardsCompatibilityInterceptor.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/BackwardsCompatibilityInterceptor.java
@@ -24,6 +24,7 @@ import okhttp3.Response;
 import okio.Buffer;
 
 import java.io.IOException;
+import java.net.HttpURLConnection;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Objects;
@@ -134,7 +135,7 @@ public class BackwardsCompatibilityInterceptor implements Interceptor {
     Response response = chain.proceed(request);
     if (isOpenshiftApiRequest(request)) {
       return handleOpenshiftRequests(request, response, chain);
-    } else if (!response.isSuccessful() && responseCodeToTransformations.keySet().contains(response.code())) {
+    } else if (!response.isSuccessful() && responseCodeToTransformations.containsKey(response.code())) {
       String url = request.url().toString();
       Matcher matcher = getMatcher(url);
       ResourceKey key = getKey(matcher);
@@ -170,7 +171,7 @@ public class BackwardsCompatibilityInterceptor implements Interceptor {
   }
 
   private static Response handleOpenshiftRequests(Request request, Response response, Chain chain) throws IOException{
-    if (!response.isSuccessful()) {
+    if (!response.isSuccessful() && response.code() == HttpURLConnection.HTTP_NOT_FOUND) {
       ResourceKey target = getResourceKeyFromRequest(request);
       if (target != null) {
         String requestUrl = request.url().toString();

--- a/kubernetes-itests/src/test/java/io/fabric8/openshift/BuildConfigIT.java
+++ b/kubernetes-itests/src/test/java/io/fabric8/openshift/BuildConfigIT.java
@@ -20,12 +20,12 @@ import io.fabric8.commons.ClusterEntity;
 import io.fabric8.commons.ReadyEntity;
 import io.fabric8.openshift.api.model.BuildConfig;
 import io.fabric8.openshift.api.model.BuildConfigList;
+import io.fabric8.openshift.api.model.BuildSourceBuilder;
 import io.fabric8.openshift.client.OpenShiftClient;
 import org.arquillian.cube.kubernetes.api.Session;
 import org.arquillian.cube.openshift.impl.requirement.RequiresOpenshift;
 import org.arquillian.cube.requirement.ArquillianConditionalRunner;
 import org.jboss.arquillian.test.api.ArquillianResource;
-import org.junit.AfterClass;
 import org.junit.BeforeClass;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -87,5 +87,24 @@ public class BuildConfigIT {
     await().atMost(30, TimeUnit.SECONDS).until(buildConfigReady);
     boolean bDeleted = client.buildConfigs().inNamespace(session.getNamespace()).withName("bc-delete").delete();
     assertTrue(bDeleted);
+  }
+
+  @Test
+  public void createOrReplace() {
+    // Given
+    BuildConfig buildConfig = client.buildConfigs().inNamespace(session.getNamespace()).withName("bc-createorreplace").get();
+
+    // When
+    buildConfig.getSpec().setSource(new BuildSourceBuilder()
+      .withNewGit()
+      .withUri("https://github.com/openshift/test2")
+      .endGit()
+      .build());
+    buildConfig = client.buildConfigs().inNamespace(session.getNamespace()).createOrReplace(buildConfig);
+
+    // Then
+    assertNotNull(buildConfig);
+    assertEquals("bc-createorreplace", buildConfig.getMetadata().getName());
+    assertEquals("https://github.com/openshift/test2", buildConfig.getSpec().getSource().getGit().getUri());
   }
 }

--- a/kubernetes-itests/src/test/java/io/fabric8/openshift/DeploymentConfigIT.java
+++ b/kubernetes-itests/src/test/java/io/fabric8/openshift/DeploymentConfigIT.java
@@ -89,4 +89,19 @@ public class DeploymentConfigIT {
     assertTrue(bDeleted);
   }
 
+  @Test
+  public void createOrReplace() {
+    // Given
+    DeploymentConfig deploymentConfig = client.deploymentConfigs().inNamespace(session.getNamespace()).withName("dc-createorreplace").get();
+
+    // When
+    deploymentConfig.getSpec().getTemplate().getSpec().getContainers().get(0).setImage("openshift/hello-openshift:v3.8");
+    deploymentConfig = client.deploymentConfigs().inNamespace(session.getNamespace()).createOrReplace(deploymentConfig);
+
+    // Then
+    assertNotNull(deploymentConfig);
+    assertEquals("dc-createorreplace", deploymentConfig.getMetadata().getName());
+    assertEquals("openshift/hello-openshift:v3.8", deploymentConfig.getSpec().getTemplate().getSpec().getContainers().get(0).getImage());
+  }
+
 }

--- a/kubernetes-itests/src/test/java/io/fabric8/openshift/ImageStreamIT.java
+++ b/kubernetes-itests/src/test/java/io/fabric8/openshift/ImageStreamIT.java
@@ -90,4 +90,19 @@ public class ImageStreamIT {
     assertTrue(bDeleted);
   }
 
+  @Test
+  public void createOrReplace() {
+    // Given
+    ImageStream imageStream = client.imageStreams().inNamespace(session.getNamespace()).withName("is-createorreplace").get();
+
+    // When
+    imageStream.getSpec().setDockerImageRepository("docker.io/openshift/ruby-centos-2");
+    imageStream = client.imageStreams().inNamespace(session.getNamespace()).createOrReplace(imageStream);
+
+    // Then
+    assertNotNull(imageStream);
+    assertEquals("is-createorreplace", imageStream.getMetadata().getName());
+    assertEquals("docker.io/openshift/ruby-centos-2", imageStream.getSpec().getDockerImageRepository());
+  }
+
 }

--- a/kubernetes-itests/src/test/java/io/fabric8/openshift/RouteIT.java
+++ b/kubernetes-itests/src/test/java/io/fabric8/openshift/RouteIT.java
@@ -30,6 +30,7 @@ import org.junit.BeforeClass;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 
+import java.util.Collections;
 import java.util.concurrent.TimeUnit;
 
 import static junit.framework.TestCase.assertNotNull;
@@ -92,6 +93,23 @@ public class RouteIT {
     await().atMost(30, TimeUnit.SECONDS).until(route1Ready);
     boolean bDeleted = client.routes().inNamespace(currentNamespace).withName("route-delete").delete();
     assertTrue(bDeleted);
+  }
+
+  @Test
+  public void createOrReplace() {
+    // Given
+    Route route = client.routes().inNamespace(currentNamespace).withName("route-createorreplace").get();
+
+    // When
+    route.getMetadata().setAnnotations(Collections.singletonMap("foo", "bar"));
+    route.getSpec().setHost("test.fabric8.io");
+    route = client.routes().inNamespace(currentNamespace).createOrReplace(route);
+
+    // Then
+    assertNotNull(route);
+    assertEquals("route-createorreplace", route.getMetadata().getName());
+    assertEquals("bar", route.getMetadata().getAnnotations().get("foo"));
+    assertEquals("test.fabric8.io", route.getSpec().getHost());
   }
 
 }

--- a/kubernetes-itests/src/test/resources/buildconfig-it.yml
+++ b/kubernetes-itests/src/test/resources/buildconfig-it.yml
@@ -130,3 +130,33 @@ spec:
     to:
       kind: "ImageStreamTag"
       name: "origin-ruby-sample:latest"
+---
+kind: "BuildConfig"
+apiVersion: "v1"
+metadata:
+  name: bc-createorreplace
+spec:
+  triggers:
+    - type: "GitHub"
+      github:
+        secret: "secret101"
+    - type: "Generic"
+      generic:
+        secret: "secret101"
+    - type: "ImageChange"
+  source:
+    type: "Git"
+    git:
+      uri: "https://github.com/openshift/ruby-hello-world"
+    dockerfile: "FROM openshift/ruby-22-centos7\nUSER example"
+  strategy:
+    type: "Source"
+    sourceStrategy:
+      from:
+        kind: "ImageStreamTag"
+        name: "ruby-20-centos7:latest"
+  output:
+    to:
+      kind: "ImageStreamTag"
+      name: "origin-ruby-sample:latest"
+

--- a/kubernetes-itests/src/test/resources/deploymentconfig-it.yml
+++ b/kubernetes-itests/src/test/resources/deploymentconfig-it.yml
@@ -142,3 +142,35 @@ spec:
           name: "origin-ruby-sample:latest"
   strategy:
     type: "Rolling"
+---
+kind: "DeploymentConfig"
+apiVersion: "v1"
+metadata:
+  name: dc-createorreplace
+spec:
+  template:
+    metadata:
+      labels:
+        name: "frontend"
+    spec:
+      containers:
+        - name: "helloworld"
+          image: "openshift/origin-ruby-sample"
+          ports:
+            - containerPort: 8080
+              protocol: "TCP"
+  replicas: 5
+  selector:
+    name: "frontend"
+  triggers:
+    - type: "ConfigChange"
+    - type: "ImageChange"
+      imageChangeParams:
+        automatic: true
+        containerNames:
+          - "helloworld"
+        from:
+          kind: "ImageStreamTag"
+          name: "origin-ruby-sample:latest"
+  strategy:
+    type: "Rolling"

--- a/kubernetes-itests/src/test/resources/imagestream-it.yml
+++ b/kubernetes-itests/src/test/resources/imagestream-it.yml
@@ -42,3 +42,10 @@ metadata:
   name: "is-delete"
 spec:
   dockerImageRepository: "docker.io/openshift/ruby-20-centos7"
+---
+apiVersion: "v1"
+kind: "ImageStream"
+metadata:
+  name: "is-createorreplace"
+spec:
+  dockerImageRepository: "docker.io/openshift/ruby-20-centos7"

--- a/kubernetes-itests/src/test/resources/route-it.yml
+++ b/kubernetes-itests/src/test/resources/route-it.yml
@@ -54,3 +54,13 @@ spec:
   to:
     kind: Service
     name: service-name
+---
+apiVersion: v1
+kind: Route
+metadata:
+  name: route-createorreplace
+spec:
+  host: www.example.com
+  to:
+    kind: Service
+    name: service-name

